src/lib.rs

```diff
@@ -0,0 +1,414 @@
+            match actions.iter().filter_map(|a| a.get_task()).next() {
+                Some(Task::Pause) => {
+                    let mut guard = self.pause.lock().unwrap();
+                    *guard = true;
+                    loop {
+                        guard = self.pause_notifier.wait(guard).unwrap();
+                        if !*guard {
+                            break;
+                        }
+                    }
+                    return None;
+                }
+                Some(t) => t,
+                None => return None,
+            }
+        };
+
+        match task {
+            Task::Off => {}
+            Task::Return(s) => return Some(s),
+            Task::Sleep(t) => thread::sleep(Duration::from_millis(t)),
+            Task::Panic(msg) => match msg {
+                Some(ref msg) => panic!("{}", msg),
+                None => panic!("failpoint {} panic", self.name),
+            },
+            Task::Print(msg) => match msg {
+                Some(ref msg) => println!("{}", msg),
+                None => println!("failpoint {} executed.", self.name),
+            },
+            Task::Pause => unreachable!(),
+            Task::Yield => thread::yield_now(),
+            Task::Delay(t) => {
+                let timer = Instant::now();
+                let timeout = Duration::from_millis(t);
+                while timer.elapsed() < timeout {}
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::sync::{mpsc, Arc};
+
+    #[test]
+    fn test_off() {
+        let point = FailPoint::new("test_fail_point_off");
+        point.set_actions(vec![Action::new(Task::Off, 1.0, None)]);
+        assert!(point.eval().is_none());
+    }
+
+    #[test]
+    fn test_return() {
+        let point = FailPoint::new("test_fail_point_return");
+        point.set_actions(vec![Action::new(Task::Return(None), 1.0, None)]);
+        let res = point.eval();
+        assert_eq!(res, Some(None));
+
+        let ret = Some("test".to_owned());
+        point.set_actions(vec![Action::new(Task::Return(ret.clone()), 1.0, None)]);
+        let res = point.eval();
+        assert_eq!(res, Some(ret));
+    }
+
+    #[test]
+    fn test_sleep() {
+        let point = FailPoint::new("test_fail_point_sleep");
+        let timer = Instant::now();
+        point.set_actions(vec![Action::new(Task::Sleep(1000), 1.0, None)]);
+        assert!(point.eval().is_none());
+        assert!(timer.elapsed() > Duration::from_millis(1000));
+    }
+
+    #[should_panic]
+    #[test]
+    fn test_panic() {
+        let point = FailPoint::new("test_fail_point_panic");
+        point.set_actions(vec![Action::new(Task::Panic(None), 1.0, None)]);
+        point.eval();
+    }
+
+    #[test]
+    fn test_print() {
+        let point = FailPoint::new("test_fail_point_print");
+        point.set_actions(vec![Action::new(Task::Print(None), 1.0, None)]);
+        assert!(point.eval().is_none());
```

Comment:
Please check stdout.

Reply:
I may change it to log latter, so let's check it then.